File: whitemagic-zig/src/memory/galactic_telepathy.zig
// Galactic Telepathy & Horizontal Gene Transfer (Zig Implementation)
//
// Q4 Cross-Database Synchronization - Zig Implementation
// Optimized for memory efficiency and deterministic performance

const std = @import("std");
const crypto = std.crypto;
const json = std.json;
const time = std.time;
const fs = std.fs;
const path = std.fs.path;

/// Types of sync conflicts
pub const SyncConflictType = enum {
    divergent_content,
    divergent_metadata,
    orphan_association,
    embedding_mismatch,
    timestamp_inversion,
};

/// Represents a sync conflict
pub const SyncConflict = struct {
    conflict_type: SyncConflictType,
    memory_id: []const u8,
    source_galaxy: []const u8,
    target_galaxy: []const u8,
    details: std.StringHashMap([]const u8),
    suggested_resolution: []const u8,
    auto_resolvable: bool,

    pub fn init(allocator: std.mem.Allocator) SyncConflict {
        return .{
            .conflict_type = .divergent_content,
            .memory_id = &.{},
            .source_galaxy = &.{},
            .target_galaxy = &.{},
            .details = std.StringHashMap([]const u8).init(allocator),
            .suggested_resolution = "manual_review",
            .auto_resolvable = false,
        };
    }

    pub fn deinit(self: *SyncConflict) void {
        self.details.deinit();
    }
};

/// Embedding bundle for transfer
pub const EmbeddingBundle = struct {
    memory_id: []const u8,
    model_name: []const u8,
    vector: []f32,
    dimensions: u32,
    created_at: []const u8,

    pub fn init(allocator: std.mem.Allocator, mem_id: []const u8, model: []const u8, dims: u32) !EmbeddingBundle {
        return .{
            .memory_id = try allocator.dupe(u8, mem_id),
            .model_name = try allocator.dupe(u8, model),
            .vector = try allocator.alloc(f32, dims),
            .dimensions = dims,
            .created_at = try allocator.dupe(u8, "1970-01-01T00:00:00Z"),
        };
    }

    pub fn deinit(self: *EmbeddingBundle, allocator: std.mem.Allocator) void {
        allocator.free(self.memory_id);
        allocator.free(self.model_name);
        allocator.free(self.vector);
        allocator.free(self.created_at);
    }
};

/// Association bundle for transfer
pub const AssociationBundle = struct {
    source_id: []const u8,
    target_id: []const u8,
    relation_type: []const u8,
    strength: f32,
    direction: []const u8,
    edge_type: ?[]const u8,

    pub fn init(allocator: std.mem.Allocator, src: []const u8, tgt: []const u8, rel: []const u8) !AssociationBundle {
        return .{
            .source_id = try allocator.dupe(u8, src),
            .target_id = try allocator.dupe(u8, tgt),
            .relation_type = try allocator.dupe(u8, rel),
            .strength = 0.5,
            .direction = try allocator.dupe(u8, "bidirectional"),
            .edge_type = null,
        };
    }

    pub fn deinit(self: *AssociationBundle, allocator: std.mem.Allocator) void {
        allocator.free(self.source_id);
        allocator.free(self.target_id);
        allocator.free(self.relation_type);
        allocator.free(self.direction);
        if (self.edge_type) |et| allocator.free(et);
    }
};

/// Memory record structure
pub const MemoryRecord = struct {
    id: []const u8,
    content: []const u8,
    content_hash: []const u8,
    title: ?[]const u8,
    memory_type: []const u8,
    tags: std.ArrayList([]const u8),
    importance: f32,
    emotional_valence: ?f32,
    metadata: std.StringHashMap([]const u8),
    created_at: f64,
    modified_at: f64,
    galactic_distance: ?f32,

    pub fn init(allocator: std.mem.Allocator, content: []const u8, mem_type: []const u8) !MemoryRecord {
        const id = try std.fmt.allocPrint(allocator, "mem_{d}", .{time.milliTimestamp()});
        const hash = try computeContentHash(allocator, content);
        const now = @as(f64, @floatFromInt(time.milliTimestamp())) / 1000.0;

        return .{
            .id = id,
            .content = try allocator.dupe(u8, content),
            .content_hash = hash,
            .title = null,
            .memory_type = try allocator.dupe(u8, mem_type),
            .tags = std.ArrayList([]const u8).init(allocator),
            .importance = 0.5,
            .emotional_valence = null,
            .metadata = std.StringHashMap([]const u8).init(allocator),
            .created_at = now,
            .modified_at = now,
            .galactic_distance = null,
        };
    }

    pub fn deinit(self: *MemoryRecord, allocator: std.mem.Allocator) void {
        allocator.free(self.id);
        allocator.free(self.content);
        allocator.free(self.content_hash);
        if (self.title) |t| allocator.free(t);
        allocator.free(self.memory_type);
        for (self.tags.items) |tag| allocator.free(tag);
        self.tags.deinit();
        var meta_it = self.metadata.iterator();
        while (meta_it.next()) |entry| {
            allocator.free(entry.key_ptr.*);
            allocator.free(entry.value_ptr.*);
        }
        self.metadata.deinit();
    }
};

/// Compute SHA-256 hash of content
pub fn computeContentHash(allocator: std.mem.Allocator, content: []const u8) ![]const u8 {
    var hash: [crypto.hash.sha2.Sha256.digest_length]u8 = undefined;
    crypto.hash.sha2.Sha256.hash(content, &hash, .{});
    
    const hex_len = hash.len * 2;
    const hex = try allocator.alloc(u8, hex_len);
    _ = try std.fmt.bufPrint(hex, "{s}", .{std.fmt.fmtSliceHexLower(&hash)});
    return hex;
}

/// Sync watermark for incremental sync
pub const SyncWatermark = struct {
    galaxy_pair: []const u8,
    last_sync_timestamp: f64,
    last_memory_id: ?[]const u8,
    memories_synced_count: u64,
    associations_synced_count: u64,
    embeddings_synced_count: u64,

    pub fn init(allocator: std.mem.Allocator, pair: []const u8) !SyncWatermark {
        return .{
            .galaxy_pair = try allocator.dupe(u8, pair),
            .last_sync_timestamp = 0,
            .last_memory_id = null,
            .memories_synced_count = 0,
            .associations_synced_count = 0,
            .embeddings_synced_count = 0,
        };
    }

    pub fn deinit(self: *SyncWatermark, allocator: std.mem.Allocator) void {
        allocator.free(self.galaxy_pair);
        if (self.last_memory_id) |id| allocator.free(id);
    }
};

/// Conflict resolution strategies
pub const ConflictResolution = enum {
    timestamp_wins,
    source_wins,
    target_wins,
    manual,
};

/// Sync options
pub const SyncOptions = struct {
    since_timestamp: ?f64,
    include_embeddings: bool,
    include_associations: bool,
    conflict_resolution: ConflictResolution,

    pub fn default() SyncOptions {
        return .{
            .since_timestamp = null,
            .include_embeddings = true,
            .include_associations = true,
            .conflict_resolution = .timestamp_wins,
        };
    }
};

/// Transfer result
pub const TransferResult = struct {
    status: []const u8,
    new_id: ?[]const u8,
    embeddings: u32,
    associations: u32,
    error: ?[]const u8,

    pub fn success(allocator: std.mem.Allocator, new_id: []const u8, embs: u32, assocs: u32) !TransferResult {
        return .{
            .status = "transferred",
            .new_id = try allocator.dupe(u8, new_id),
            .embeddings = embs,
            .associations = assocs,
            .error = null,
        };
    }

    pub fn failure(error_msg: []const u8) TransferResult {
        return .{
            .status = "error",
            .new_id = null,
            .embeddings = 0,
            .associations = 0,
            .error = error_msg,
        };
    }

    pub fn deinit(self: *TransferResult, allocator: std.mem.Allocator) void {
        if (self.new_id) |id| allocator.free(id);
        if (self.error) |err| allocator.free(err);
    }
};

/// Sync result
pub const SyncResult = struct {
    source: []const u8,
    target: []const u8,
    since_timestamp: f64,
    candidates: u64,
    transferred: u64,
    skipped: u64,
    conflicts: u64,
    embeddings_transferred: u64,
    associations_transferred: u64,
};

/// Main telepathy engine
pub const GalacticTelepathyEngine = struct {
    allocator: std.mem.Allocator,
    watermarks: std.StringHashMap(SyncWatermark),
    watermark_path: []const u8,
    mutex: std.Thread.Mutex,

    pub fn init(allocator: std.mem.Allocator) !GalacticTelepathyEngine {
        const home_dir = try std.process.getEnvVarOwned(allocator, "HOME");
        defer allocator.free(home_dir);
        
        const wm_path = try path.join(allocator, &[_][]const u8{ home_dir, ".whitemagic", "sync_registry.json" });
        
        var engine = GalacticTelepathyEngine{
            .allocator = allocator,
            .watermarks = std.StringHashMap(SyncWatermark).init(allocator),
            .watermark_path = wm_path,
            .mutex = .{},
        };
        
        try engine.loadWatermarks();
        return engine;
    }

    pub fn deinit(self: *GalacticTelepathyEngine) void {
        var it = self.watermarks.iterator();
        while (it.next()) |entry| {
            entry.value_ptr.deinit(self.allocator);
        }
        self.watermarks.deinit();
        self.allocator.free(self.watermark_path);
    }

    fn loadWatermarks(self: *GalacticTelepathyEngine) !void {
        const file = fs.cwd().openFile(self.watermark_path, .{}) catch |err| switch (err) {
            error.FileNotFound => return, // No watermarks yet
            else => |e| return e,
        };
        defer file.close();

        const content = try file.readToEndAlloc(self.allocator, 1024 * 1024);
        defer self.allocator.free(content);

        // Parse JSON (simplified - would need proper JSON parser)
        // For now, watermarks start empty
    }

    fn saveWatermarks(self: *GalacticTelepathyEngine) !void {
        self.mutex.lock();
        defer self.mutex.unlock();

        // Ensure directory exists
        const dir_path = path.dirname(self.watermark_path) orelse return error.InvalidPath;
        fs.cwd().makePath(dir_path) catch {};

        // Serialize and save (simplified implementation)
        const file = try fs.cwd().createFile(self.watermark_path, .{});
        defer file.close();

        // Write JSON representation
        try file.writeAll("{\"watermarks\": {}}");
    }

    fn getWatermarkKey(source: []const u8, target: []const u8) []const u8 {
        // Returns canonical ordering
        // In real implementation: return if source < target then source:target else target:source
        return source;
    }

    /// Perform incremental sync between galaxies
    pub fn incrementalSync(
        self: *GalacticTelepathyEngine,
        source_galaxy: []const u8,
        target_galaxy: []const u8,
        opts: SyncOptions,
    ) !SyncResult {
        _ = self;
        _ = source_galaxy;
        _ = target_galaxy;
        _ = opts;
        
        // Placeholder implementation
        return SyncResult{
            .source = source_galaxy,
            .target = target_galaxy,
            .since_timestamp = 0,
            .candidates = 0,
            .transferred = 0,
            .skipped = 0,
            .conflicts = 0,
            .embeddings_transferred = 0,
            .associations_transferred = 0,
        };
    }

    /// Detect conflicts
    fn detectConflict(
        self: *GalacticTelepathyEngine,
        mem: *const MemoryRecord,
    ) ?SyncConflict {
        _ = self;
        _ = mem;
        return null;
    }

    /// Resolve conflicts
    fn resolveConflict(
        self: *GalacticTelepathyEngine,
        conflict: *SyncConflict,
        strategy: ConflictResolution,
    ) void {
        _ = self;
        _ = conflict;
        _ = strategy;
    }

    /// Multi-hop federated sync
    pub fn federatedSync(
        self: *GalacticTelepathyEngine,
        galaxy_chain: [][]const u8,
        opts: SyncOptions,
    ) !void {
        _ = self;
        _ = galaxy_chain;
        _ = opts;
    }
};

/// Test function for compilation verification
pub fn testCompilation() void {
    std.debug.print("Galactic Telepathy Zig module compiled successfully\n", .{});
}

/// FFI exports for C interop
export fn telepathy_engine_create() ?*GalacticTelepathyEngine {
    const allocator = std.heap.page_allocator;
    const engine = allocator.create(GalacticTelepathyEngine) catch return null;
    engine.* = GalacticTelepathyEngine.init(allocator) catch {
        allocator.destroy(engine);
        return null;
    };
    return engine;
}

export fn telepathy_engine_destroy(engine: ?*GalacticTelepathyEngine) void {
    if (engine) |e| {
        e.deinit();
        std.heap.page_allocator.destroy(e);
    }
}

export fn telepathy_incremental_sync(
    engine: ?*GalacticTelepathyEngine,
    source: [*c]const u8,
    target: [*c]const u8,
) c_int {
    if (engine == null or source == null or target == null) return -1;
    
    const src = std.mem.span(source);
    const tgt = std.mem.span(target);
    
    const opts = SyncOptions.default();
    _ = engine.?.incrementalSync(src, tgt, opts) catch return -1;
    
    return 0;
}

/// Main entry for testing
pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    testCompilation();

    var engine = try GalacticTelepathyEngine.init(allocator);
    defer engine.deinit();

    const opts = SyncOptions.default();
    const result = try engine.incrementalSync("default", "archive", opts);

    std.debug.print("Sync result: transferred {d} memories\n", .{result.transferred});
}

File: whitemagic-go/galactic_telepathy.go
// Galactic Telepathy & Horizontal Gene Transfer (Go Implementation)
//
// Q4 Cross-Database Synchronization - Go Implementation
// Optimized for concurrent, high-throughput memory transfer
// with goroutine-based parallel processing.

package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"sync"
	"time"
)

// SyncConflictType represents types of conflicts during transfer
type SyncConflictType string

const (
	DivergentContent    SyncConflictType = "divergent_content"
	DivergentMetadata   SyncConflictType = "divergent_metadata"
	OrphanAssociation   SyncConflictType = "orphan_association"
	EmbeddingMismatch   SyncConflictType = "embedding_mismatch"
	TimestampInversion  SyncConflictType = "timestamp_inversion"
)

// SyncConflict represents a conflict requiring resolution
type SyncConflict struct {
	ConflictType        SyncConflictType `json:"conflict_type"`
	MemoryID            string           `json:"memory_id"`
	SourceGalaxy        string           `json:"source_galaxy"`
	TargetGalaxy        string           `json:"target_galaxy"`
	Details             map[string]interface{} `json:"details"`
	SuggestedResolution string           `json:"suggested_resolution"`
	AutoResolvable      bool             `json:"auto_resolvable"`
}

// EmbeddingBundle holds embedding data for transfer
type EmbeddingBundle struct {
	MemoryID   string    `json:"memory_id"`
	ModelName  string    `json:"model_name"`
	Vector     []float32 `json:"vector"`
	Dimensions uint32    `json:"dimensions"`
	CreatedAt  string    `json:"created_at"`
}

// AssociationBundle holds association data for transfer
type AssociationBundle struct {
	SourceID     string                 `json:"source_id"`
	TargetID     string                 `json:"target_id"`
	RelationType string                 `json:"relation_type"`
	Strength     float32                `json:"strength"`
	Direction    string                 `json:"direction"`
	EdgeType     *string                `json:"edge_type,omitempty"`
	Metadata     map[string]interface{} `json:"metadata"`
}

// MemoryRecord represents a memory in the system
type MemoryRecord struct {
	ID               string                 `json:"id"`
	Content          string                 `json:"content"`
	ContentHash      string                 `json:"content_hash"`
	Title            *string                `json:"title,omitempty"`
	MemoryType       string                 `json:"memory_type"`
	Tags             []string               `json:"tags"`
	Importance       float32                `json:"importance"`
	EmotionalValence *float32               `json:"emotional_valence,omitempty"`
	Metadata         map[string]interface{} `json:"metadata"`
	CreatedAt        float64                `json:"created_at"`
	ModifiedAt       float64                `json:"modified_at"`
	GalacticDistance *float32               `json:"galactic_distance,omitempty"`
}

// ComputeContentHash calculates SHA-256 hash of content
func ComputeContentHash(content string) string {
	hash := sha256.Sum256([]byte(content))
	return hex.EncodeToString(hash[:])
}

// NewMemoryRecord creates a new memory record with computed hash
func NewMemoryRecord(content string, title *string, memoryType string) MemoryRecord {
	now := float64(time.Now().Unix())
	return MemoryRecord{
		ID:          fmt.Sprintf("mem_%d", time.Now().UnixNano()),
		Content:     content,
		ContentHash: ComputeContentHash(content),
		Title:       title,
		MemoryType:  memoryType,
		Tags:        []string{},
		Importance:  0.5,
		Metadata:    make(map[string]interface{}),
		CreatedAt:   now,
		ModifiedAt:  now,
	}
}

// SyncWatermark tracks last sync point
type SyncWatermark struct {
	GalaxyPair              string  `json:"galaxy_pair"`
	LastSyncTimestamp       float64 `json:"last_sync_timestamp"`
	LastMemoryID            *string `json:"last_memory_id,omitempty"`
	MemoriesSyncedCount     uint64  `json:"memories_synced_count"`
	AssociationsSyncedCount uint64  `json:"associations_synced_count"`
	EmbeddingsSyncedCount   uint64  `json:"embeddings_synced_count"`
}

// ConflictResolution strategy enum
type ConflictResolution int

const (
	TimestampWins ConflictResolution = iota
	SourceWins
	TargetWins
	Manual
)

// SyncOptions configures sync behavior
type SyncOptions struct {
	SinceTimestamp       *float64
	IncludeEmbeddings    bool
	IncludeAssociations  bool
	ConflictResolution ConflictResolution
}

// DefaultSyncOptions returns default options
func DefaultSyncOptions() SyncOptions {
	return SyncOptions{
		IncludeEmbeddings:    true,
		IncludeAssociations:  true,
		ConflictResolution:   TimestampWins,
	}
}

// SyncResult summarizes a sync operation
type SyncResult struct {
	Source                 string `json:"source"`
	Target                 string `json:"target"`
	SinceTimestamp         float64 `json:"since_timestamp"`
	Candidates             uint64 `json:"candidates"`
	Transferred            uint64 `json:"transferred"`
	Skipped               uint64 `json:"skipped"`
	Conflicts             uint64 `json:"conflicts"`
	EmbeddingsTransferred uint64 `json:"embeddings_transferred"`
	AssociationsTransferred uint64 `json:"associations_transferred"`
}

// HopResult is a single hop in federated sync
type HopResult struct {
	Source      string `json:"source"`
	Target      string `json:"target"`
	Memories    uint64 `json:"memories"`
	Associations uint64 `json:"associations"`
	Embeddings  uint64 `json:"embeddings"`
}

// FederationResult is the result of multi-hop sync
type FederationResult struct {
	Chain             []string    `json:"chain"`
	Hops              []HopResult `json:"hops"`
	TotalMemories     uint64      `json:"total_memories"`
	TotalAssociations uint64      `json:"total_associations"`
	TotalEmbeddings   uint64      `json:"total_embeddings"`
}

// MemoryBackend interface for storage operations
type MemoryBackend interface {
	GetMemory(id string) (*MemoryRecord, error)
	StoreMemory(record *MemoryRecord) (string, error)
	FindByContentHash(hash string) (*string, error)
	GetModifiedSince(timestamp float64) ([]MemoryRecord, error)
	GetEmbeddings(memoryID string) ([]EmbeddingBundle, error)
	StoreEmbedding(memoryID string, embedding *EmbeddingBundle) error
	GetAssociations(memoryID string) ([]AssociationBundle, error)
	StoreAssociation(assoc *AssociationBundle) error
	GetStats() map[string]uint64
	GetGalaxyName() string
}

// GalaxyManager interface
type GalaxyManager interface {
	GetMemoryBackend(galaxyName string) (MemoryBackend, error)
	ListGalaxies() []string
}

// GalacticTelepathyEngine is the main sync engine
type GalacticTelepathyEngine struct {
	watermarks     map[string]SyncWatermark
	watermarkPath  string
	galaxyManager  GalaxyManager
	mu             sync.RWMutex
}

// NewGalacticTelepathyEngine creates a new engine
func NewGalacticTelepathyEngine(gm GalaxyManager) *GalacticTelepathyEngine {
	homeDir, _ := os.UserHomeDir()
	watermarkPath := filepath.Join(homeDir, ".whitemagic", "sync_registry.json")
	
	engine := &GalacticTelepathyEngine{
		watermarks:    make(map[string]SyncWatermark),
		watermarkPath: watermarkPath,
		galaxyManager: gm,
	}
	
	engine.loadWatermarks()
	return engine
}

// loadWatermarks loads from disk
func (e *GalacticTelepathyEngine) loadWatermarks() {
	data, err := os.ReadFile(e.watermarkPath)
	if err != nil {
		return // File doesn't exist yet
	}
	
	var loaded map[string]SyncWatermark
	if err := json.Unmarshal(data, &loaded); err != nil {
		fmt.Fprintf(os.Stderr, "Failed to parse watermarks: %v\n", err)
		return
	}
	
	e.watermarks = loaded
}

// saveWatermarks persists to disk
func (e *GalacticTelepathyEngine) saveWatermarks() error {
	e.mu.RLock()
	defer e.mu.RUnlock()
	
	data, err := json.MarshalIndent(e.watermarks, "", "  ")
	if err != nil {
		return err
	}
	
	dir := filepath.Dir(e.watermarkPath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return err
	}
	
	return os.WriteFile(e.watermarkPath, data, 0644)
}

// getWatermarkKey creates canonical key for galaxy pair
func getWatermarkKey(source, target string) string {
	if source < target {
		return fmt.Sprintf("%s:%s", source, target)
	}
	return fmt.Sprintf("%s:%s", target, source)
}

// IncrementalSync performs sync only for changed memories
func (e *GalacticTelepathyEngine) IncrementalSync(
	sourceGalaxy,
	targetGalaxy string,
	opts SyncOptions,
) (*SyncResult, error) {
	wmKey := getWatermarkKey(sourceGalaxy, targetGalaxy)
	
	// Determine since timestamp
	var since float64
	if opts.SinceTimestamp != nil {
		since = *opts.SinceTimestamp
	} else if wm, ok := e.watermarks[wmKey]; ok {
		since = wm.LastSyncTimestamp
	}
	
	// Get backends
	srcBackend, err := e.galaxyManager.GetMemoryBackend(sourceGalaxy)
	if err != nil {
		return nil, fmt.Errorf("source galaxy not found: %w", err)
	}
	
	tgtBackend, err := e.galaxyManager.GetMemoryBackend(targetGalaxy)
	if err != nil {
		return nil, fmt.Errorf("target galaxy not found: %w", err)
	}
	
	// Get modified memories
	candidates, err := srcBackend.GetModifiedSince(since)
	if err != nil {
		return nil, fmt.Errorf("failed to get modified memories: %w", err)
	}
	
	result := &SyncResult{
		Source:         sourceGalaxy,
		Target:         targetGalaxy,
		SinceTimestamp: since,
		Candidates:     uint64(len(candidates)),
	}
	
	// Process transfers concurrently
	var wg sync.WaitGroup
	semaphore := make(chan struct{}, 10) // Limit concurrent transfers
	
	var mu sync.Mutex
	
	for _, mem := range candidates {
		wg.Add(1)
		semaphore <- struct{}{}
		
		go func(m MemoryRecord) {
			defer wg.Done()
			defer func() { <-semaphore }()
			
			transferResult, err := e.transferMemoryWithFidelity(
				&m, srcBackend, tgtBackend, opts,
			)
			
			mu.Lock()
			defer mu.Unlock()
			
			if err != nil {
				result.Conflicts++
				return
			}
			
			if transferResult.Status == "transferred" {
				result.Transferred++
				result.EmbeddingsTransferred += uint64(transferResult.Embeddings)
				result.AssociationsTransferred += uint64(transferResult.Associations)
			} else {
				result.Skipped++
			}
		}(mem)
	}
	
	wg.Wait()
	
	// Update watermark
	e.mu.Lock()
	now := float64(time.Now().Unix())
	var lastID *string
	if len(candidates) > 0 {
		id := candidates[len(candidates)-1].ID
		lastID = &id
	}
	
	e.watermarks[wmKey] = SyncWatermark{
		GalaxyPair:              wmKey,
		LastSyncTimestamp:       now,
		LastMemoryID:            lastID,
		MemoriesSyncedCount:     result.Transferred,
		AssociationsSyncedCount: result.AssociationsTransferred,
		EmbeddingsSyncedCount:   result.EmbeddingsTransferred,
	}
	e.mu.Unlock()
	
	// Save watermarks
	if err := e.saveWatermarks(); err != nil {
		fmt.Fprintf(os.Stderr, "Failed to save watermarks: %v\n", err)
	}
	
	return result, nil
}

// TransferResult represents the outcome of a single memory transfer
type TransferResult struct {
	Status       string  `json:"status"`
	NewID        *string `json:"new_id,omitempty"`
	Embeddings   uint32  `json:"embeddings"`
	Associations uint32  `json:"associations"`
	Error        *string `json:"error,omitempty"`
}

// transferMemoryWithFidelity transfers a memory with full context
func (e *GalacticTelepathyEngine) transferMemoryWithFidelity(
	mem *MemoryRecord,
	srcBackend, tgtBackend MemoryBackend,
	opts SyncOptions,
) (*TransferResult, error) {
	// Check for conflicts
	conflict := e.detectConflict(mem, tgtBackend)
	if conflict != nil {
		switch opts.ConflictResolution {
		case Manual:
			errMsg := fmt.Sprintf("Manual resolution required: %v", conflict)
			return &TransferResult{
				Status: "conflict",
				Error:  &errMsg,
			}, nil
		default:
			e.resolveConflict(conflict, opts.ConflictResolution, tgtBackend)
		}
	}
	
	// Extract embeddings
	var embeddings []EmbeddingBundle
	if opts.IncludeEmbeddings {
		embeddings, _ = srcBackend.GetEmbeddings(mem.ID)
	}
	
	// Extract associations
	var associations []AssociationBundle
	if opts.IncludeAssociations {
		associations, _ = srcBackend.GetAssociations(mem.ID)
	}
	
	// Create new memory record
	newMem := *mem
	newMem.ID = fmt.Sprintf("mem_%d", time.Now().UnixNano())
	newMem.Tags = append(newMem.Tags, fmt.Sprintf("hgt_source:%s", srcBackend.GetGalaxyName()))
	
	// Update metadata
	now := float64(time.Now().Unix())
	newMem.Metadata["telepathy_source"] = srcBackend.GetGalaxyName()
	newMem.Metadata["telepathy_source_id"] = mem.ID
	newMem.Metadata["telepathy_timestamp"] = now
	newMem.Metadata["telepathy_version"] = "2.0-go"
	
	// Store in target
	newID, err := tgtBackend.StoreMemory(&newMem)
	if err != nil {
		return nil, err
	}
	
	// Restore embeddings
	embCount := 0
	for _, emb := range embeddings {
		emb.MemoryID = newID
		if err := tgtBackend.StoreEmbedding(newID, &emb); err == nil {
			embCount++
		}
	}
	
	// Restore associations
	assocCount := 0
	for _, assoc := range associations {
		// Remap IDs
		if assoc.SourceID == mem.ID {
			assoc.SourceID = newID
		}
		if assoc.TargetID == mem.ID {
			assoc.TargetID = newID
		}
		
		if err := tgtBackend.StoreAssociation(&assoc); err == nil {
			assocCount++
		}
	}
	
	return &TransferResult{
		Status:       "transferred",
		NewID:        &newID,
		Embeddings:   uint32(embCount),
		Associations: uint32(assocCount),
	}, nil
}

// detectConflict checks for potential sync conflicts
func (e *GalacticTelepathyEngine) detectConflict(
	mem *MemoryRecord,
	tgtBackend MemoryBackend,
) *SyncConflict {
	// Check if memory with same content hash exists
	existingID, err := tgtBackend.FindByContentHash(mem.ContentHash)
	if err != nil || existingID == nil {
		return nil
	}
	
	// Check if it's from same source
	existing, err := tgtBackend.GetMemory(*existingID)
	if err != nil || existing == nil {
		return nil
	}
	
	if sourceID, ok := existing.Metadata["telepathy_source_id"].(string); ok {
		if sourceID == mem.ID {
			// Same memory, already synced
			return nil
		}
	}
	
	// Divergent content with same hash (collision)
	if existing.Content != mem.Content {
		return &SyncConflict{
			ConflictType:        DivergentContent,
			MemoryID:            mem.ID,
			SourceGalaxy:        "unknown",
			TargetGalaxy:        "unknown",
			Details: map[string]interface{}{
				"target_existing_id": *existingID,
				"content_hash":       mem.ContentHash,
			},
			SuggestedResolution: "manual_review",
			AutoResolvable:      false,
		}
	}
	
	return nil
}

// resolveConflict applies conflict resolution strategy
func (e *GalacticTelepathyEngine) resolveConflict(
	conflict *SyncConflict,
	strategy ConflictResolution,
	tgtBackend MemoryBackend,
) {
	// Implementation depends on strategy
	switch strategy {
	case TimestampWins:
		// Keep newer (handled by insertion logic)
	case SourceWins:
		// Overwrite with source
	case TargetWins:
		// Keep target (skip transfer)
	}
}

// FederatedSync performs multi-hop synchronization
func (e *GalacticTelepathyEngine) FederatedSync(
	galaxyChain []string,
	opts SyncOptions,
) (*FederationResult, error) {
	if len(galaxyChain) < 2 {
		return nil, fmt.Errorf("chain requires at least 2 galaxies")
	}
	
	result := &FederationResult{
		Chain: galaxyChain,
	}
	
	for i := 0; i < len(galaxyChain)-1; i++ {
		source := galaxyChain[i]
		target := galaxyChain[i+1]
		
		hopResult, err := e.IncrementalSync(source, target, opts)
		if err != nil {
			return nil, fmt.Errorf("hop %d (%s->%s) failed: %w", i, source, target, err)
		}
		
		result.Hops = append(result.Hops, HopResult{
			Source:       source,
			Target:       target,
			Memories:     hopResult.Transferred,
			Associations: hopResult.AssociationsTransferred,
			Embeddings:   hopResult.EmbeddingsTransferred,
		})
		
		result.TotalMemories += hopResult.Transferred
		result.TotalAssociations += hopResult.AssociationsTransferred
		result.TotalEmbeddings += hopResult.EmbeddingsTransferred
	}
	
	return result, nil
}

// GetWatermark returns current watermark for a galaxy pair
func (e *GalacticTelepathyEngine) GetWatermark(source, target string) (*SyncWatermark, bool) {
	e.mu.RLock()
	defer e.mu.RUnlock()
	
	wm, ok := e.watermarks[getWatermarkKey(source, target)]
	if !ok {
		return nil, false
	}
	return &wm, true
}

// ResetWatermark clears watermark for a galaxy pair (force full sync)
func (e *GalacticTelepathyEngine) ResetWatermark(source, target string) error {
	e.mu.Lock()
	defer e.mu.Unlock()
	
	delete(e.watermarks, getWatermarkKey(source, target))
	return e.saveWatermarks()
}

// ListWatermarks returns all tracked watermarks
func (e *GalacticTelepathyEngine) ListWatermarks() []string {
	e.mu.RLock()
	defer e.mu.RUnlock()
	
	keys := make([]string, 0, len(e.watermarks))
	for k := range e.watermarks {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}
